compiler/rustc_middle/src/hir/map/mod.rs

```diff
@@ -929,27 +929,34 @@ impl<'hir> Map<'hir> {
         }
     }
 
+    pub(super) fn opt_ident_span(self, id: HirId) -> Option<Span> {
+        let ident = match self.get(id) {
+            // A `Ctor` doesn't have an identifier itself, but its parent
+            // struct/variant does. Compare with `hir::Map::opt_span`.
+            Node::Ctor(..) => match self.find(self.get_parent_node(id))? {
+                Node::Item(item) => Some(item.ident),
+                Node::Variant(variant) => Some(variant.ident),
+                _ => unreachable!(),
+            },
+            node => node.ident(),
+        };
+        ident.map(|ident| ident.span)
+    }
+
     pub fn opt_name(self, id: HirId) -> Option<Symbol> {
-        Some(match self.get(id) {
-            Node::Item(i) => i.ident.name,
-            Node::ForeignItem(fi) => fi.ident.name,
-            Node::ImplItem(ii) => ii.ident.name,
-            Node::TraitItem(ti) => ti.ident.name,
-            Node::Variant(v) => v.ident.name,
-            Node::Field(f) => f.ident.name,
-            Node::Lifetime(lt) => lt.name.ident().name,
-            Node::GenericParam(param) => param.name.ident().name,
-            Node::Binding(&Pat { kind: PatKind::Binding(_, _, l, _), .. }) => l.name,
-            Node::Ctor(..) => self.name(HirId::make_owner(self.get_parent_item(id))),
-            _ => return None,
-        })
+        match self.get(id) {
+            Node::Binding(&Pat { kind: PatKind::Binding(_, _, l, _), .. }) => Some(l.name),
+            Node::Ctor(..) => match self.find(self.get_parent_node(id))? {
+                Node::Item(item) => Some(item.ident.name),
+                Node::Variant(variant) => Some(variant.ident.name),
+                _ => unreachable!(),
+            },
+            node => node.ident().map(|i| i.name),
```

Comment:
Maybe there should be a second "smart" `ident()` function including both `Node::Binding` and `Node::Ctor`, then both `fn opt_ident_span` and `fn opt_name` could be implemented through it.